File: Utils/src/hkccpacmanrobot/utils/map/MapKey.java
package hkccpacmanrobot.utils.map;

/**
 * Created by 13058456a on 3/18/2015.
 */
public class MapKey implements Cloneable{
    public final long x, y;

    public MapKey(long x, long y) {
        this.x = x;
        this.y = y;
    }

    public MapKey(double x, double y) {
        this.x = Math.round(x / 5);
        this.y = Math.round(y / 5);
    }

    @Override
    protected Object clone() {
        return new MapKey(x,y);
    }
}

File: src/main/scala/edu/hkcc/pacmanrobot/utils/map/MapKey.scala
package edu.hkcc.pacmanrobot.utils.map

/**
 * Created by beenotung on 3/21/15.
 */
class MapKey(val x: Int, val y: Int) extends Cloneable with Serializable {
  override def clone: AnyRef = {
    new MapKey(x, y)
  }

  override def equals(o: Any): Boolean = {
    if (o.isInstanceOf[MapKey])
      equals(o.asInstanceOf[MapKey])
    else false
  }

  def equals(mapKey: MapKey): Boolean = {
    x.equals(mapKey.x) && y.equals(mapKey.y)
  }

  override def toString: String = {
    "(" + x + "," + y + ")"
  }
}
